compat/mingw.c

```diff
@@ -475,6 +475,42 @@ static wchar_t *normalize_ntpath(wchar_t *wbuf)
 	return wbuf;
 }
 
+/* rename and unlink a file. This is needed for the ability to recreate a file
+ * with the same name, for example on checkout which unlinks and recreates the file */
+static int rename_and_unlink(const wchar_t *pathname, int (*unlinker)(const wchar_t *))
+{
+	int res;
+	wchar_t temppath[MAX_LONG_PATH];
+	*temppath = 0;
+	/* leave some space for the .unlink_XXXXXX extension */
+	wcsncat(temppath, pathname, sizeof(temppath) - 15);
+	wcsncat(temppath, L".unlink_XXXXXX", sizeof(temppath));
+	if (!_wmktemp(temppath) || !MoveFileExW(pathname, temppath, 0))
+		return 1;
+	res = unlinker(temppath);
+	if (res) {
+		const DWORD err = GetLastError();
+		MoveFileExW(temppath, pathname, 0);
+		SetLastError(err);
+	}
+	return res;
+}
+
+static int safe_unlink(const wchar_t *wpathname, int (*unlinker)(const wchar_t *))
```

Comment:
I refactored it a bit more, but I can't separate the rename from unlink.

We have to retry the renaming too, because the rename itself can fail. Just like chmod is inside the retry loop.

If unlink fails, I rename back to the original name, so the retry starts from the same state.

I have no idea how to create such a test. You need to have multiple threads/processes with exact delays to trigger this.

Reply:
I regularly encounter the retry use case when I try to do git clean on a repository while I have a solution opened in Visual Studio. I assume that it creates some files with dwShareMode 0 which "Prevents other processes from opening a file or device if they request delete, read, or write access." (https://docs.microsoft.com/en-us/windows/desktop/api/fileapi/nf-fileapi-createfilea)

It should not be difficult to write a test for this. I see no need for multiple threads/processes or exact delays to trigger this. Just open the file beforehand, then run git checkout or git clean, just like we did for the other tests. The difference is that we probably cannot use exec to open files in this way. Maybe we need to write a test-helper in C to get this behavior.

When we have a test for the retry behavior, I think we will find that retry for renames is not needed. Even if we need to keep the retry for renames for some reason, it is absolutely fine to have separate retry loops for renaming and the unlinking.

For me, the "renaming back so that we re-rename later" approach is unnecessary and just wrong.

Reply:
I can try to find a way to trigger the retry and fail every time, but I doubt I'll be able to create a test that *succeeds* on retry because it requires accurate timing.

Regarding the rename issue, I need to keep the handle until unlink is done.

I don't find much sense in retrying to rename 5 times, because if you failed on the first attempt it means that another process has an unshared handle to this file, so you won't succeed on the following attempts. But if this other process released its handle, you can continue *without* renaming.

Reply:
> I don't find much sense in retrying to rename 5 times

We do have code for that in Git itself, as it seems to help with, say, anti-malware scanners that hold a handle to the file just for a short time.

Reply:
I meant retrying *only* the rename, and then a separate loop for retrying the delete after successful rename.

I do understand the reason for retry in general.

Reply:
You don't need accurate timing to test the retry:
```
open_file_exclusively
git checkout <input_pipe >output_pipe &
while read -r line <output_pipe; do
  if $line == "retry?"; then
    close_file_exclusively
    echo y >input_pipe
  fi
done
```

Maybe it is also sufficient to test for the case where we retry and fail. Check that the file which we failed to remove still exists after failing, and that the retry code was triggered (we can see that from the output).

You have me confused regarding your stance on the rename issue.
- What is the issue with keeping the handle until unlink is done?
- I don't understand why you bring up a rename-retry-loop as an issue. The code structure which I suggested allows to either do a rename-retry-loop or not do a rename-retry-loop. It is entirely up to you. If it does not make sense to retry the rename, then just don't. If we determined that a rename is necessary, and the rename fails (with or without retries), then I would not expect to even proceed to the unlink. We cannot sensibly continue and abort right there.